noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,191 @@
+        nft_exists: Map<Field, PublicMutable<bool>>,
+        // A map from token ID to the public owner of the NFT.
+        public_owners: Map<Field, PublicMutable<AztecAddress>>,
+        // Contains the NFTs that are waiting to be shielded.
+        pending_shields: PrivateSet<TransparentNote>,
+    }
+
+    #[aztec(public)]
+    #[aztec(initializer)]
+    fn constructor(admin: AztecAddress, name: str<31>, symbol: str<31>, decimals: u8) {
+        assert(!admin.is_zero(), "invalid admin");
+        storage.admin.write(admin);
+        storage.minters.at(admin).write(true);
+        storage.name.initialize(FieldCompressedString::from_string(name));
+        storage.symbol.initialize(FieldCompressedString::from_string(symbol));
+    }
+
+    #[aztec(public)]
+    fn set_admin(new_admin: AztecAddress) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.admin.write(new_admin);
+    }
+
+    #[aztec(public)]
+    fn set_minter(minter: AztecAddress, approve: bool) {
+        assert(storage.admin.read().eq(context.msg_sender()), "caller is not admin");
+        storage.minters.at(minter).write(approve);
+    }
+
+    #[aztec(public)]
+    fn mint_public(to: AztecAddress, token_id: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
+        // assert the token_id does not exist yet and mint
+        ...
+    }
+
+    #[aztec(public)]
+    fn mint_private(token_id: Field, secret_hash: Field) {
+        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
+        let pending_shields = storage.pending_shields;
+        let mut note = TransparentNote::new(token_id, secret_hash);
+        pending_shields.insert_from_public(&mut note);
+        // set that the nft exists
+        ...
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_name() -> pub FieldCompressedString {
+        storage.name.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_name() -> pub FieldCompressedString {
+        storage.name.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn admin() -> Field {
+        storage.admin.read().to_field()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn is_minter(minter: AztecAddress) -> bool {
+        storage.minters.at(minter).read()
+    }
+
+    // Returns the total number of NFTs a given address owns in public.
+    #[aztec(public)]
+    #[aztec(view)]
+    fn balance_of_public(owner: AztecAddress) -> Field {
+        storage.public_balances.at(owner).read().to_integer()
+    }
```

Comment:
Doesn't cost us much to add this, but I'm quite unsure as to whether someone would ever call this. I think this was only added for compat with erc20 (idk why), and to support the `IERC721Enumerable` extension.

Reply:
It also seemed weird to me as by definition the quantity of NFTs you own does not tell you much due to nonfungibility of the tokens. I included it just because it's common in Ethereum's world. I am fine with nuking it.

Reply:
I'm tempted to remove this initially as part of the initiative mentioned above of just having private state.

Reply:
Up you mentioned:
> I'm fine with minting in public initially, but yeah using shared mutable for the minters seems reasonable.

which seems to be in conflict with

> the initiative mentioned above of just having private state

WDYM here?

Reply:
Nuked in https://github.com/AztecProtocol/aztec-packages/pull/8129/commits/c548b46f5974999220df2e7761f1b9a792182ebf